### PR TITLE
feat: Add global redis for registry cache

### LINKED ISSUE #34: Set up a shared redis for the proxy nodes in Terraform.

### PR DESCRIPTION
**Pull request by xav**
# Description

Add a redis instance for caching registry requests.

Resolves #34 

## Due Diligence

* [ ] Breaking change
* [ ] Requires a documentation update
* [ ] Requires a e2e/integration test update

### CODE REVIEW

File: terraform/redis_replication/main.tf [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+resource "aws_elasticache_replication_group" "cache" {
+  replication_group_id          = replace("${var.app_name}-${var.redis_name}_replica", "_", "-")
+  replication_group_description = replace("${var.app_name}-${var.redis_name}_replica", "_", "-")
+  engine                        = var.global ? "redis" : null
+  node_type                     = var.global ? var.node_type : null
+  parameter_group_name          = var.global ? "default.redis6.x" : null
+  engine_version                = var.global ? "6.x" : null
+  port                          = 6379
+  subnet_group_name             = aws_elasticache_subnet_group.private_subnets.name
```
Comment (arein) [inline terraform/redis_replication/main.tf:30]: we don't need this redis cluster to be globally replicated. When we regionalize it's simpler to just have one per region without global replication.
In the relay we have other state that actually needs to be replicated per region so it was convenient and already set up to do the same for the project cache.

File: terraform/redis_replication/main.tf [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+resource "aws_route53_record" "dns" {
+  count   = var.global ? 1 : 0
+  zone_id = aws_route53_zone.private_zone[0].id
+  name    = "${replace("${var.redis_name}-redis", "_", "-")}.${aws_route53_zone.private_zone[0].name}"
+  type    = "CNAME"
+  ttl     = "30"
+  records = [aws_elasticache_replication_group.cache.primary_endpoint_address]
+}
```
Comment (arein) [inline terraform/redis_replication/main.tf:95]: i know you're just copying this from the relay but nitpick can we not have one private hosted zone per redis? Can you set up a hosted zone for the RPC proxy instead? ideally you do this in the `infra` repository as the hosted zone will be shared when we regionalize the RPC

**Review by arein — CHANGES_REQUESTED** [review]
Good stuff! Few changes requested

Do you know why `plan` is failing?

**Review by arein — COMMENTED** [review]
How did you test this?

**Comment by xav** [comment]
> How did you test this?

This one is just for the infra, so `tf plan` basically.

**Review by arein — APPROVED** [review]

**Comment by arein** [comment]
> > How did you test this?
> 
> This one is just for the infra, so `tf plan` basically.

We have `dev` and `staging` environments. Please feel free to test your changes on one or the other (ideally `dev`). The README has instructions.

**Review by arein — APPROVED** [review]
can you make a follow up task pls to also monitor Redis CPU/memory? We have an example in the relay monitoring logic

File: terraform/monitoring/main.tf [hunk 3]
```diff
@@ -10,12 +10,12 @@ terraform {
...
-    var.environment == "prod" ?
-    "[{\"uid\": \"${local.opsgenie_notification_channel}\"}]" :
-    "[]"
-  )
+  #  opsgenie_notification_channel = "l_iaPw6nk"
+  #  notifications = (
+  #    var.environment == "prod" ?
+  #    "[{\"uid\": \"${local.opsgenie_notification_channel}\"}]" :
+  #    "[]"
+  #  )
```
[the remark this answers is not in the document]
Comment (xav) [reply terraform/monitoring/main.tf:18]: I think it's been there a while but, without linting, it was invisible.
I'll remove it altogether.

**Comment by xav** [comment]
> > > How did you test this?
> > 
> > 
> > This one is just for the infra, so `tf plan` basically.
> 
> We have `dev` and `staging` environments. Please feel free to test your changes on one or the other (ideally `dev`). The README has instructions.

Good thing you wanted to test it more thoroughly because the apply eventually failed.
I changed how the ECS containers are declared so that the JSON syntax can be checked instead of an opaque heredoc.

**Pull request merged**